src/plugins/nvdimm.c

```diff
@@ -0,0 +1,549 @@
+    }
+
+    ndctl_unref (ctx);
+    return NULL;
+}
+
+/**
+ * bd_nvdimm_list_namespaces:
+ * @bus_name: (allow-none): return only namespaces on given bus (specified by name),
+ *                          %NULL may be specified to return namespaces from all buses
+ * @region_name: (allow-none): return only namespaces on given region (specified by 'regionX' name),
+ *                             %NULL may be specified to return namespaces from all regions
+ * @idle: whether to list idle (not enabled) namespaces too
+ * @extra: (allow-none) (array zero-terminated=1): extra options (currently unused)
+ * @error: (out): place to store error (if any)
+ *
+ * Returns: (array zero-terminated=1): information about the namespaces on @bus and @region or
+ *                                     %NULL if no namespaces were found (@error may be set to indicate error)
+ *
+ * Tech category: %BD_NVDIMM_TECH_NAMESPACE-%BD_NVDIMM_TECH_MODE_QUERY
+ */
+BDNVDIMMNamespaceInfo** bd_nvdimm_list_namespaces (const gchar *bus_name, const gchar *region_name, gboolean idle, const BDExtraArg **extra UNUSED, GError **error) {
+    struct ndctl_ctx *ctx = NULL;
+    struct ndctl_namespace *ndns = NULL;
+    struct ndctl_region *region = NULL;
+    struct ndctl_bus *bus = NULL;
+    guint i = 0;
+    gint ret = 0;
+
+    BDNVDIMMNamespaceInfo **infos = NULL;
+    GPtrArray *namespaces = g_ptr_array_new ();
+
+    ret = ndctl_new (&ctx);
+    if (ret != 0) {
+        g_set_error (error, BD_NVDIMM_ERROR, BD_NVDIMM_ERROR_NAMESPACE_FAIL,
+                     "Failed to create ndctl context");
+        return NULL;
+    }
+
+    ndctl_bus_foreach (ctx, bus) {
+        if (bus_name && g_strcmp0 (bus_name, ndctl_bus_get_devname (bus)) != 0)
+            continue;
+
+        ndctl_region_foreach (bus, region) {
+            if (region_name && g_strcmp0 (bus_name, ndctl_region_get_devname (region)) != 0)
+                continue;
+
+            ndctl_namespace_foreach (region, ndns) {
+                if (!idle && !ndctl_namespace_is_active (ndns))
+                    continue;
+
+                BDNVDIMMNamespaceInfo *info = get_nvdimm_namespace_info (ndns, error);
+                if (!info) {
+                    g_ptr_array_foreach (namespaces, (GFunc) (void *) bd_nvdimm_namespace_info_free, NULL);
+                    g_ptr_array_free (namespaces, FALSE);
+                    ndctl_unref (ctx);
+                    return NULL;
+                }
+
+                g_ptr_array_add (namespaces, info);
+            }
+        }
+    }
+
+    if (namespaces->len == 0) {
+        ndctl_unref (ctx);
+        return NULL;
+    }
+
+    /* now create the return value -- NULL-terminated array of BDNVDIMMNamespaceInfo */
+    infos = g_new0 (BDNVDIMMNamespaceInfo*, namespaces->len + 1);
+    for (i=0; i < namespaces->len; i++)
+        infos[i] = (BDNVDIMMNamespaceInfo*) g_ptr_array_index (namespaces, i);
+    infos[i] = NULL;
+
+    g_ptr_array_free (namespaces, FALSE);
```

Comment:
With ``g_ptr_array_free (namespaces, FALSE)`` there's no need to create the return value with the ``for`` cycle above, the inner contents of the ``GPointerArray`` can be used as is.